### PR TITLE
Add a NOOPT macro to try and block optimisations in tests.

### PR DESCRIPTION
**Pull request by vext01**
**This is a draft PR for discussion for now. I'll adapt the existing tests if this is the right approach.**

For context, see:
https://stackoverflow.com/questions/37786547/enforcing-statement-order-in-c/38025837#38025837

We use (the latest version of) the method proposed there.

In the new test, I checked that without using NOOPT, the compiler was
able to optimise `x = 2; y = 3; res = x + y` to `res = 5`. And that with
NOOPT, the addition remains intact.

### CODE REVIEW

File: c_tests/tests/noopts.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <yk_testing.h>
+
+int add_them(int x, int y) { return x + y; }
+
+int main(int argc, char **argv) {
+  int x, y, res;
+  void *tt = __yktrace_start_tracing(HW_TRACING, &res, &x, &y);
+  // The __yktrace_start_tracing() call has almost certainly already blocked
```
Comment (ltratt) [inline c_tests/tests/noopts.c:26]: This paragraph isn't necessarily true *if* `__yktrace_start_tracing` can be inlined. I'm not sure the comment is very helpful TBH.

File: ykcapi/yk_testing.h [hunk 2]
```diff
@@ -20,3 +20,10 @@ void __yktrace_irtrace_get(void *trace, size_t idx, char **res_func,
                            size_t *res_bb);
 void *__yktrace_irtrace_compile(void *trace);
 void __yktrace_drop_irtrace(void *trace);
+
+// Force sequentialisation of instructions surrounding a computation.
```
Comment (ltratt) [inline ykcapi/yk_testing.h:24]: This is hard to parse, and it's not correct, because we can't stop *all* computations being moved around: we can just stop *some* computations being moved around (see e.g. http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4455.html). Rather it's preventing the compiler optimising away this variable at this point *and* putting a sort-of barrier that prevents computation on this variable being moved around.

[Also we can't "borrow" code without copyright. But I don't think we're borrowing from Google here. We're just implementing the same well-known technique -- one can find this technique used in many places. FWIW, I think the SO post is a better explanation than the header file.]

File: c_tests/tests/noopts.c [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  // optimisation of x and y, as pointers to them are passed as arguments and
+  // the compiler must assume that it may mutate them. Therefore we initialise
+  // x and y after the call.
+  x = 2;
+  NOOPT(x);
+  y = 3;
+  NOOPT(y);
+  // An optimising compiler would inline this call and (without our NOOPTs)
+  // would then use constant prorogation to replace the whole call with the
```
Comment (ltratt) [inline c_tests/tests/noopts.c:35]: I wasn't expecting parliament to make its effects felt here :) s/prorogation/propogation/

File: c_tests/tests/noopts.c [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  // value 5. We check that this doesn't happen.
+  res = add_them(x, y);
```
Comment (ltratt) [inline c_tests/tests/noopts.c:37]: I am not convinced that `add_them` can't be inlined: I am convinced that the `asm` trick prevents this turning into a constant however.

File: c_tests/tests/noopts.c (hunk 1 above)
Comment (vext01) [reply c_tests/tests/noopts.c:26]: `__yktrace_start_tracing()` is in an external shared object, so it can't be inlined.

The reason I included this comment is because the first version of the test initialised `x` and `y` outside of the traced section of code and the compiler was already unable (without `NOOPT`) to replace the call with a constant due to `__yktrace_start_tracing`.

Comment (ltratt) [reply c_tests/tests/noopts.c:26]: OK. I think I'd just remove the comment to be honest.

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (vext01) [reply ykcapi/yk_testing.h:24]: > preventing the compiler optimising away this variable at this point.

That doesn't capture what's going on here either.

How about "Tries to trick the compiler into thinking that the asm block may mutate the value, thus blocking certain optimisations". This is my understanding.

> one can find this technique used in many places

The problem is that, although it is an old trick, there are lots of different variants of it.

There's the one from the stack overflow post:
```
asm volatile("" : "+m"(const_cast<T &>(value)));
```

which is not used any more it seems.

Then variations in the current google benchmark tree are as follows:

 * The const variant:  `asm volatile("" : : "r,m"(value) : "memory");`
 * The clang non-const variant: `asm volatile("" : "+r,m"(value) : : "memory");`
 * The non-clang non-const variant: `asm volatile("" : "+m,r"(value) : : "memory");`

And then I've seen other variants too like
 * `asm("":::"memory");` -- the one you see a lot in C circles.
 * Rust's black box: `llvm_asm!("" : : "r"(&mut dummy) : "memory" : "volatile");` (equiv to `asm volatile("": "r"(value), "memory");` I think) which is yet another variant.

So actually I think citing the source of our block is quite important here.

File: c_tests/tests/noopts.c (hunk 1 above)
Comment (vext01) [reply c_tests/tests/noopts.c:26]: I don't often push back, but I think this is a critical comment.

Someone may be tempted to move the initialisation up outside of the traced section, at which point we no longer test `NOOPT`. At that point we'd be testing if `__yk_start_tracing()` blocks optimisations, which is not the intention of the test.

Comment (ltratt) [reply c_tests/tests/noopts.c:26]: Surely all I have to do is something like:

```c
int x = 1;
start_tracing(&x);
x = 2;
NOOPT(x);
...
```

? In other words, I don't think *initialisation* is very special per se.

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (ltratt) [reply ykcapi/yk_testing.h:24]: > How about "Tries to trick the compiler into thinking that the asm block may mutate the value, thus blocking certain optimisations".

This is closer, but note that we're only preventing optimisations *on that variable* (or an expression if you pass in an expression). In other words, this has sort-of local effects on a variable, not sort-of global effects on every part of the function.

> So actually I think citing the source of our block is quite important here.

Dumb question: why are we using this rather than one of the others? Is it a bit of a random guess?

File: c_tests/tests/noopts.c (hunk 1 above)
Comment (vext01) [reply c_tests/tests/noopts.c:26]: OK, you got me.

s/initialisation/assignment/

We can't expect the compiler to constant fold an *assignment* that happened before the call to `start_tracing()`, as it doesn't know if it was mutated.

What I don't want is for someone to change the test to the (shorter):
```
int x = 1, y = 2;
NOOPT(X);
NOOPT(Y);
start_tracing()
...
```

As here, whether or not `NOOPT` blocks optimisations is moot, as we know that (currently) `start_tracing` already blocks optimisations. Thus we wouldn't be testing what we thought we were. `NOOPT` could be `#define NOOPT(x) x` and the test would still pass.

Does that make sense?

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (vext01) [reply ykcapi/yk_testing.h:24]: I chose the google non-const clang variant, because:
 - we use clang
 - the value may not be constant.
 - the google implementations are likely to be correct (?)

What do you reckon?

File: c_tests/tests/noopts.c (hunk 3 above)
Comment (vext01) [reply c_tests/tests/noopts.c:35]: Just you wait until you make a typo ;)

File: c_tests/tests/noopts.c (hunk 4 above)
Comment (vext01) [reply c_tests/tests/noopts.c:37]: You are right, inlining is irrelevant.

File: c_tests/tests/noopts.c (hunk 1 above)
Comment (ltratt) [reply c_tests/tests/noopts.c:26]: I now realise that (once again!) there are two levels of IR we're thinking about here: `NOOPT` is meant to inhibit optimisations in both the AOT and JIT IRs. [I must get used to thinking at both levels!] I think our test is meant to test only that it does so in the JIT IR? If that is the case (and it's probably worth being clear about that in a comment), then it's obvious that the "IR part of the test" can only happen between stop/start tracing.

Comment (vext01) [reply c_tests/tests/noopts.c:26]: I think here we'd be checking that it doesn't happen in the AOTMod, since `YKD_PRINT_IR` prints the `IR` before it's submitted to the execution engine.

TBH, I don't care where we test if `NOOPT` works, I just want to know that it does work!

Comment (ltratt) [reply c_tests/tests/noopts.c:26]: In our case I think we do want to know that it works at both IR levels! It would be confusing if it worked in the AOT IR but not the JIT IR. It wouldn't hurt to test both IRs, but at the moment we only have the JIT IR debugging as output -- so we might as well test that. In fairness, if it works in the JIT IR, it almost certainly works in the AOT IR!

Comment (vext01) [reply c_tests/tests/noopts.c:26]: OK. Testing the post-opt JIT module would require a larger change (and an additional test). 

Let's put this on hold until @ptersilie can figure out the blockmap bug.

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (ltratt) [reply ykcapi/yk_testing.h:24]: So the "real" reference is actually the talk at https://youtu.be/nXaxk27zwlk?t=2441 -- it's worth watching! He defines `escape` and `clobber` which correspond to `DoNotOptimize` (which due to C++ overloading comes in several flavours) and `ClobberMemory` in the "Google code". Note that they have both:

```c++
inline BENCHMARK_ALWAYS_INLINE void DoNotOptimize(Tp const& value)
```

and:

```c++
inline BENCHMARK_ALWAYS_INLINE void DoNotOptimize(Tp& value) {
```

I am not sure why they need to overload `const&`, but @vext01, your better knowledge of x86 asm might help here?

**Comment by ltratt** [comment]
One minor thing that occurred to me: I think we should guard this with something like `#if defined(LLVM)` just to emphasise that we've only tested this with LLVM. Just in case someone comes along one day, ports all the other LLVM stuff to (say) the MS compiler and then this bit silently doesn't quite do the right thing.

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (vext01) [reply ykcapi/yk_testing.h:24]: In light of new info -- that in a commit to Google benchmark, the definition shown in the talk was deemed buggy -- what shall we do with this?

I still think I'd like to cite Google. They've clearly though hard about this.

**Comment by vext01** [comment]
> just to emphasise that we've only tested this with LLVM.

Is it necessary? The whole approach is underpinned by LLVM IR :P

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (ltratt) [reply ykcapi/yk_testing.h:24]: > I still think I'd like to cite Google. They've clearly though hard about this.

OK.

Do we have an answer for the `const&` thing?

**Comment by ltratt** [comment]
I think this is more general than just this repo, and I wouldn't be surprised if at some future point someone takes this stuff out and tries it in other compilers. Guarding it for LLVM costs us little, but will help our future selves.

**Comment by vext01** [comment]
OK, but it's the clients of this repo that need to be built with ykllvm, so such guards would need to go in them, not us.

In theory you could build yk with gcc and it would still work so long as the client use clang.

**Comment by ltratt** [comment]
The compile-time guards need to go around our `NOOPT` stuff.

**Comment by vext01** [comment]
I'm adding the GCC variant for the reason above :)

**Comment by ltratt** [comment]
Let's not add what we've not tested.

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (vext01) [reply ykcapi/yk_testing.h:24]: > In other words, this has sort-of local effects on a variable, not sort-of global effects on every part of the function.

Notice that the whole memory space is still clobbered.

So in fact, the `clobber` operation is weaker than the `escape` operation. I didn't expect that.

So how about:

> Blocks the compiler from optimising the specified value or expression.

**Comment by vext01** [comment]
89c6d67bd9 addresses comments and allows us to test all of the various IR stages in the pipeline. It looks worse than it is because I had to move some functions around.

I've changed our environment variables so that you can do things like:
```
YKD_PRINT_IR=aot,jit-post-opt
```

Which would print the IR at those two stages in the pipeline.

Each IR print has a header which can then be anchored in the lang test to check that specific IR.

Only the NOOPT test has been updated to use this new facility. If you like what you see, I will update the others.

**Comment by vext01** [comment]
@ptersilie you might like to take a look too.

File: c_tests/tests/noopts.c [hunk 5]
```diff
@@ -18,22 +40,24 @@
 #include <string.h>
 #include <yk_testing.h>
 
-int add_them(int x, int y) { return x + y; }
+int add_them(int x, int y) {
+  // At the call-site x and y are constant, so an optimising compiler would try
```
Comment (ltratt) [inline c_tests/tests/noopts.c:44]: Putting the comment here feels weird given that this isn't where the noopt effect happens.

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (ltratt) [reply ykcapi/yk_testing.h]: I'm not going to let go of the `const` question so let's start with that :)

**Comment by ptersilie** [comment]
Looks good. Do we know why the `ExecutionEngine` doesn't optimise even when given an opt level via the `EngineBuilder`?

File: ykllvmwrap/src/ykllvmwrap.cc [hunk 6]
```diff
@@ -62,6 +64,146 @@ thread_local bool ThreadAOTModInitialized = false;
...
+// Dumps an LLVM Value to a string and trims leading whitespace.
+void dumpValueToString(Value *V, string &S) {
+  raw_string_ostream RSO(S);
+  V->print(RSO);
+  lTrim(S, ' ');
+}
+#endif
+
+enum DebugIR {
+  AOT,
```
Comment (ltratt) [inline ykllvmwrap/src/ykllvmwrap.cc:82]: Is it guaranteed that these are given the numbers 0,1,2,3?

Comment (vext01) [reply ykllvmwrap/src/ykllvmwrap.cc:82]: I believe so:

>  If the first enumerator does not have an initializer, the associated value is zero. For any other enumerator whose definition does not have an initializer, the associated value is the value of the previous enumerator plus one. 

https://en.cppreference.com/w/cpp/language/enum

**Comment by vext01** [comment]
>  Do we know why the ExecutionEngine doesn't optimise even when given an opt level via the EngineBuilder?

Afraid not. I could dig further, but I'm starting to grow tired of this PR :P

File: c_tests/tests/noopts.c (hunk 5 above)
Comment (vext01) [reply c_tests/tests/noopts.c:44]: But it is where the optimisation is blocked. Removal of the plus operation is blocked.

File: ykllvmwrap/src/ykllvmwrap.cc (hunk 6 above)
Comment (ltratt) [reply ykllvmwrap/src/ykllvmwrap.cc:82]: Perfect!

File: c_tests/tests/noopts.c (hunk 5 above)
Comment (ltratt) [reply c_tests/tests/noopts.c:44]: Sorry I wasn't very clear. I don't think this function needs to be broken out. I think it's equivalent to something like:

```
x = NOOPT(...);
y = NOOPT(...);
z = x + y;
```

i.e. breaking the function out isn't necessary (and is confusing).

File: ykcapi/yk_testing.h (hunk 2 above)
Comment (vext01) [reply ykcapi/yk_testing.h]: Comparing the const and non-const version:
```
    const: asm volatile("" :               : "r,m"(value) : "memory");
non-const: asm volatile("" : "+r,m"(value) :              : "memory");
```

And the syntax is:


Where the meaning of the input and output strings is:
>      OutputOperands:  A comma-separated list of the C variables modified by the instructions in the AssemblerTemplate. An empty list is permitted. See OutputOperands.
InputOperands
>
>    A comma-separated list of C expressions read by the instructions in the AssemblerTemplate. An empty list is permitted. See InputOperands.

And the `+` (allowed only for output operands) means:
> Output constraints must begin with either ‘=’ (a variable overwriting an existing value) or ‘+’ (when reading and writing).

So putting this all together, the const version asks the compiler to put put the constant value in either memory or a register and informs that the block **reads** the constant.

The non-const version asks the compiler to do the same, but tells the compiler that the block **both reads and writes** to the variable.

My guess is that it's like this because you can't write to a constant.

Comment (vext01) [reply ykcapi/yk_testing.h]: This begs the question of whether we need a constant version.

I get the impression that this isn't an exact science, and that these funky asm blocks are just maximising the chances of hosing optimisation (with no absolute guarantee).

Note that the non-const version when applied to a constant in our test, **did* block optimisation.

I want to move on. What would you like to do?

Comment (ltratt) [reply ykcapi/yk_testing.h]: I'm happy with moving on and a) not having the const variant b) but documenting that it could exist (so that we know if/when we need it).

Comment (vext01) [reply ykcapi/yk_testing.h]: OK. 3e476be30 addresses that.

**Comment by vext01** [comment]
So @ltratt, I take it the new IR printing idea was OK? If so I will now update the existing tests.

**Comment by ltratt** [comment]
> I take it the new IR printing idea was OK?

Works for me!

**Comment by vext01** [comment]
@ltratt see those last commits.

The second one is a bit frustrating. See the commit message. I'd be interested in your thoughts.

I'm also now questioning the correctness of `NOOPT`, as I just realised that both Carruth and Google always work on a pointer (or a reference) to the data they want to block optimisations on. We are working directly on the values.

So whilst our current `NOOPT` sems to work, it not precisely what Google do in their benchmark repo.

I'm going to try and fix that but it requires more fiddling because:

 a) It can't be done with hygene in a C macro.
 b) our tracer bombs out if it encounters the obvious alternative: a `static inline` function.

Bear with me...

**Comment by ltratt** [comment]
You're right, according to the video at least, you have to do things like:

```c
x = ...;
NOOPT(&x);
```

in order to keep `x` on the stack.

**Comment by ltratt** [comment]
I'm not sure what the problem in the second commit is?

**Comment by vext01** [comment]
> you have to do things like NOOPT(&x)

Indeed, but inline asm won't take `&` to constraints, so we'd have to use an intermediate variable, and you can't do that hygienically. I'm trying to make it a function, but the trace compiler is now choking on something.

> I'm not sure what the problem in the second commit is?

Instead of a lengthy discussion, let me ask: are you OK using `black_box()` to block optimisations in rust code?

**Comment by ltratt** [comment]
> Indeed, but inline asm won't take & to constraints, so we'd have to use an intermediate variable, and you can't do that hygienically with a macro.

To be honest, I don't understand what this means!

> are you OK using black_box() to block optimisations in rust code?

Yes, I assumed we would use `black_box` rather than trying to invent our own thing that does the same!

**Comment by vext01** [comment]
> Yes, I assumed we would use black_box

OK, so there is no problem there at least. I was trying to translate the C definition to Rust, which it turns out is quite hard.

**Comment by vext01** [comment]
So now, silly question: why not expose `black_box()` to C via the yk testing API?

**Comment by ltratt** [comment]
>  So now, silly question: why not expose black_box() to C via the yk testing API?

Good idea! Why not tyr?

**Comment by vext01** [comment]
@ltratt @lukasd @jhughes The plot thickens yet further. This is going to be a bit of a write-up I'm afraid. And it's full of nitty-gritty detail. Brace yourself and I'll try my best to explain what I think is going on.

Consider we trace this:
```
  int res;
  void *tt = __yktrace_start_tracing(HW_TRACING, &res);
  res = 0;
  NOOPT(&res);
  f(&res);
  void *tr = __yktrace_stop_tracing(tt);
```

And that `NOOPT` is now a function accepting a pointer, as discussed offline (clang checks elided for clarity):
```
static inline void NOOPT(void *x) {
  asm volatile("" : "+r,m"(x) : : "memory");
}
```

The AOT IR then looks like this:
```
%5 = bitcast i32* %3 to i8*, !dbg !44
...
%6 = call i8* (i64, ...) @__yktrace_start_tracing(i64 1, i32* nonnull %3) #7, !dbg !45
...
call fastcc void @NOOPT(i8* nonnull %5), !dbg !46
...
%7 = call i8* @__yktrace_stop_tracing(i8* %6) #7, !dbg !50
...
```

This is problematic, as the generation of the pointer to `res` has been moved outside of the traced section, despite the use of `NOOPT`. This means that we end up with an undefined variable (`%5`) in our pre-opt JIT code. We need the bitcast operation to stay in side the traced section.

Why can this happen? Due to `NOOPT(&res)`, the compiler must assume that the pointer itself (`&res`) may be mutated **and** that the pointee (`*&res`) too may be mutated. But note that:

 - Mutations to the pointer itself would be local to the frame of `NOOPT()`, as the pointer is passed by value, and thus copied in.
 - Mutating the pointee does not alter its address on the stack at the call-site of `NOOPT`.

This gives the compiler license to move the computation of `&res` up and out of the traced section, as the definition of `NOOPT` can't possibly alter the address of `res` on the stack at the call-site of NOOPT.

In other words, we have blocked optimisations to `res`, but not to `&res`. By introducing `&res` we introduce a second value that we'd need to block optimisations on. But you can't just do `NOOPT(&&res)` as then you'd just make a third thing (`&&res`) that you'd then need to block optimisations on, and so on.

[This also shows that the call to black_box() in start_tracing() is not sufficient to block that particular optimisation either]

What is interesting though, is that if we use our old definition of `NOOPT`:

```
#define NOOPT(X) asm volatile("" : "+r,m"(X) : : "memory");
```

And pass the `NOOPT()` arguments by value, not pointer (and not that there is no actual call, as it's a macro function -- the asm block is just inlined), then the problem goes away. It goes away because there is no longer any "address of" operation, and thus no bitcast to "move up".

The question remains though, does doing `NOOPT(x)` -- i.e. not `NOOPT(&x)` -- do what we intend.

What I can say with full confidence is:

 - a) passing a value to the macro version of NOOPT macro does not do quite the same as Google's implementation of `DoNotOptimize`. I disassembled a usage ofDoNotOptimise to check that a pointer is passed at the binary level. It is. That means that they too would block optimisations on `res`, but not `&res`.

 - b) passing a value `x` to the macro version of NOOPT does block optimisations on `x` (today on a Linux system).

Once you've understood a), you realise that our requirements are *stronger* than Google's, and that's why they can get away with passing a pointer.
They require only that the **pointee** isn't optimised; we require than neither the **pointee** nor the **pointer** be optimised. We can't block optimisations to the **pointer**, but we can work around the problem by ensuring that no pointer exists to call NOOPT.

What's actionable then? Given b), I suggest we use the current macro version of NOOPT and let tests guide us. If at any point our NOOPT test fails, we that we got it wrong, and that we have to try again.

 I hope this makes sense. It's been very hard to explain.

**Comment by vext01** [comment]
[I realise that the example isn't the best -- will try to update]

**Comment by vext01** [comment]
> [I realise that the example isn't the best -- will try to update]

Updated, please refresh.

**Comment by ptersilie** [comment]
> Lukasd

You pinged some random dude here. ;-)

**Comment by vext01** [comment]
> We can't block optimisations to the pointer, but we can work around the problem by ensuring that no pointer exists to call NOOPT.

I now realise that this is a more general problem that extends beyond the scope of `NOOPT`.

If the user wants to trace a call to any function that accepts a pointer argument, and that call is in the top-level frame directly between start_tracing() and stop_tracing(), and LLVM deems that a bitcast is required, then will hit this same problem: the bitcast can be hoisted out of the traced section.

**Comment by vext01** [comment]
> You pinged some random dude here. ;-)

Oops, sorry!

**Comment by ptersilie** [comment]
We can't even easily unhoist the bitcast back inbetween start/stoptracing, because the arguments we pass to `start_tracing` still reference the pre-bitcast value.

However! We've always planned to automate the passing values into the trace part, i.e. we automatically figure out which values are needed inside the trace, but were defined outside of it. In that case we would then figure out, that `%7` is actually needed, and not `%3` and automatically pass in `%7`. Wouldn't that solve this issue?

**Comment by jacob-hughes** [comment]
> > Lukasd
> 
> You pinged some random dude here. ;-)

Same here. Nothing to see here Jon ;)

**Comment by vext01** [comment]
> Wouldn't that solve this issue?

Lukas is right. This would solve the issue.

OK, so here's the facts:
 - Our NOOPT macro does block the optmisations that we want, even if it's different to Google's implementation.
 - Google's implementation would require a function that accepts a pointer.
 - In general we can't inline functions that accept pointers without implementing the auto-trace-inputs Lukas mentions.
 - auto-trace-inputs is no small undertaking: it likely requires all manner of low-level juggling including stack maps and a fair chunk of inline asm.

Let's implement an imperfect solution using our current `NOOPT` and adjust it later when we have auto-trace inputs.

I'll let this settle, and update the PR on Monday if all is still well.

**Comment by ltratt** [comment]
I'm fine with moving on, with one suggestion: if our `NOOPT` works on values (not pointers) let's call it `NOOPT_VAL` (or similar) so that if we want a pointer version later we don't have to disentangle it.

**Comment by vext01** [comment]
> NOOPT_VAL

Good idea.

**Comment by vext01** [comment]
3005599f2 does the rename and some other tidy-ups. A separate commit updates the gitignore file.

OK to squash?

File: c_tests/tests/noopts.c [hunk 7]
```diff
@@ -7,67 +7,52 @@
...
-  int x, y, res;
-  void *tt = __yktrace_start_tracing(HW_TRACING, &res, &x, &y);
-  // The __yktrace_start_tracing() call has almost certainly already blocked
-  // optimisation of x and y, as pointers to them are passed as arguments and
-  // the compiler must assume that it may mutate them. Therefore to properly
-  // test NOOPT we need initialise x and y after the call.
+  int x, res;
+  void *tt = __yktrace_start_tracing(HW_TRACING, &res, &x);
+  // __yktrace_start_tracing() will already inhibit optimisations on x since
+  // it's address is passed as an argument and the compiler must assume that it
```
Comment (ltratt) [inline c_tests/tests/noopts.c:41]: s/it's/its/

File: c_tests/tests/noopts.c [hunk 8]
```diff
@@ -7,67 +7,52 @@
...
+  // may be mutated. Therefore to properly test NOOPT_VAL we need initialise x
```
Comment (ltratt) [inline c_tests/tests/noopts.c:42]: s/we need initialise/we need to initialise/

[I also think that backticks around the variable names would make this easier to read!]

File: c_tests/tests/noopts.c [hunk 9]
```diff
@@ -7,67 +7,52 @@
...
+  // and y after the call to __yktrace_start_tracing..
   x = 2;
-  NOOPT(&x);
-  y = 3;
-  NOOPT(&y);
-  res = add_them(x, y);
+  NOOPT_VAL(x);
+  res = x + 3; // We don't want this operation to be optimised away.
```
Comment (ltratt) [inline c_tests/tests/noopts.c:46]: I claim that `NOOPT_VAL(res)` is not necessary in this example because in `res = x + 3`, `x` must be opaque.

File: ykcapi/yk_testing.h [hunk 10]
```diff
@@ -23,16 +23,16 @@ void __yktrace_drop_irtrace(void *trace);
...
+// This is similar to the non-const variant borrowed from Google benchmark:
 // https://github.com/google/benchmark/blob/e451e50e9b8af453f076dec10bd6890847f1624e/include/benchmark/benchmark.h#L350
 //
-// Note that Google benchmark also defines a variant of this for constant data.
-// At the time of writing, the below NOOPT seems to suffice (even for
-// constants), but we may need to borrow the const version later.
-static inline void NOOPT(void *x) {
+// Our version works on a value, rather than a pointer.
+//
+// Note that Google benchmark also defines a variant for constant data. At the
```
Comment (ltratt) [inline ykcapi/yk_testing.h:31]: s/note that Google/note that the Google/

File: ykcapi/yk_testing.h [hunk 11]
```diff
@@ -23,16 +23,16 @@ void __yktrace_drop_irtrace(void *trace);
...
+// time of writing, the below NOOPT_VAL seems to suffice (even for constants),
```
Comment (ltratt) [inline ykcapi/yk_testing.h:32]: s/the below//

File: ykcapi/yk_testing.h [hunk 12]
```diff
@@ -23,16 +23,16 @@ void __yktrace_drop_irtrace(void *trace);
...
+// but we may need to borrow the const version later.
```
Comment (ltratt) [inline ykcapi/yk_testing.h:33]: s/borrow/consider using/

File: c_tests/tests/noopts.c (hunk 9 above)
Comment (vext01) [reply c_tests/tests/noopts.c:46]: It's just convention to guard the inputs and the outputs of the computation you want to shield from optimisations:

> The key to these techniques is to focus on the data. You make the input to the computation opaque to the optimizer and the result of the computation opaque to the optimizer.

https://stackoverflow.com/questions/11748362/how-to-disable-an-optimization-in-llvm

I don't mind either way, it's not strictly required for this particular test on Linux today.

File: ykcapi/yk_testing.h (hunk 10 above)
Comment (vext01) [reply ykcapi/yk_testing.h:31]: For better or worse, "Google benchmark" is the name of the project.

https://github.com/google/benchmark

File: c_tests/tests/noopts.c (hunk 9 above)
Comment (ltratt) [reply c_tests/tests/noopts.c:46]: I'm not sure that SO link is the right one?

Comment (vext01) [reply c_tests/tests/noopts.c:46]: Oops! https://stackoverflow.com/questions/37786547/enforcing-statement-order-in-c/38025837#38025837

**Comment by vext01** [comment]
That last commit should address all of the comments?

File: ykcapi/yk_testing.h (hunk 10 above)
Comment (ltratt) [reply ykcapi/yk_testing.h:31]: Then it should presumably be "Google Benchmark" (capital 'b')?

Comment (vext01) [reply ykcapi/yk_testing.h]: 72060fb

File: c_tests/tests/noopts.c (hunk 9 above)
Comment (ltratt) [reply c_tests/tests/noopts.c:46]: You're right and I'm wrong: it's not that we need the second `NOOPT_VAL` *as such* but putting it before `stop_tracing` ensures that `res` won't float past `stop_tracing`.

[FWIW, `res` seems a less than ideal variable name in many of these tests. It would probably be better just to use obviously meaningless names like `y`. Maybe we can think about that in a future PR?]

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.
